src/Data/Quaternion.purs

```diff
@@ -128,9 +128,14 @@ instance ringQuaternion :: Ring a => Ring (Quaternion a) where
   sub (Quaternion a1 b1 c1 d1) (Quaternion a2 b2 c2 d2) =
     Quaternion (a1 - a2) (b1 - b2) (c1 - c2) (d1 - d2)
 
+
+instance functorQuaternion :: Functor Quaternion where
+  map f (Quaternion a b c d) = Quaternion (f a) (f b) (f c) (f d)
```

Comment:
Immediately, for dual quaternions using Dual numbers as the coefficients of a Quaternion. You would like to be able to reach into the Quaternion and conjugate its coefficients, or extract only the real components, etc. Along the same lines but more generally: conversion between Quaternions with coefficients of different types.
Some of the existing arithmetic operations in Quaternion.purs may be recast in terms of map, which may be better or worse depending on your perspective.